File: Graph/PHP/Graph.php
<?php

class Graph
{
    public array $graph = [];
    public array $visited = [];


    public function insert(int $a, int $b): void
    {
        if (!isset($this->visited[$a])) {
            $this->visited[$a] = false;
        }

        if (!isset($this->visited[$b])) {
            $this->visited[$b] = false;
        }

        if (isset($this->graph[$a])) {
            $this->graph[$a][] = $b;
        }
        if (!isset($this->graph[$a])) {
            $this->graph[$a] = [$b];
        }
        if (isset($this->graph[$b])) {
            $this->graph[$b][] = $a;
        }
        if (!isset($this->graph[$b])) {
            $this->graph[$b] = [$a];
        }
    }

    /**
     * Breath First Search(BFS)
     */
    public function bfs($start): void
    {
        $this->visited[$start] = true;

        $queue = [$start];

        while (count($queue) > 0) {

            echo $queue[0] . "\n";

            foreach ($this->graph[$queue[0]] as $node) {

                if (isset($this->visited[$node]) && !$this->visited[$node]) {
                    $this->visited[$node] = true;
                    $queue[] = $node;
                }

            }

            array_shift($queue);

        }
    }
}


$graph = new Graph();

$graph->insert(10, 15);
$graph->insert(10, 20);
$graph->insert(10, 25);
$graph->insert(10, 30);
$graph->insert(15, 20);
$graph->insert(15, 25);
$graph->insert(40, 50);


//$graph->bfs(10);
//var_dump($graph->visited);
$count = 0;
foreach ($graph->visited as $key => $visited) {

    if (!$graph->visited[$key]) {

        $graph->bfs($key);

        $count++;

    }

}

echo $count;

File: Graph/PYTHON/Graph.py
class Graph:

    def __init__(self):
        self.graph = {}


    def add_edge(self,u,v):
        
        if u not in self.graph:
            self.graph[u]=[]

        if v not in self.graph:
            self.graph[v]=[]

        self.graph[u].append(v)
        self.graph[v].append(u)


    # Breadth-First Search (BFS)

    def bfs(self, start):
        visited = set()
        queue= [start]
        visited.add(start)

        while queue:

            node = queue.pop(0)
            print(node)

            for neighbor  in self.graph[node]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)



    # depth-First Search (DFS)
    def dfs(self,start):
        visited = set()
        
        def dfs_recursive(node):
            visited.add(node)
            print(node)
            for neighbor in self.graph[node]:
                if neighbor not in visited:
                    dfs_recursive(neighbor)
        dfs_recursive(start)
    

    # Cycle Detection
    def has_cycle(self):
        visited = set()

        def dfs_cycle(node, parent):
            visited.add(node)
            for neighbor in self.graph[node]:
                if neighbor not in visited:
                    if dfs_cycle(neighbor, node):
                        return True
                elif neighbor != parent:
                    return True
            return False
                
        for vertex in self.graph:
            if vertex not in visited:
                if dfs_cycle(vertex, None):
                    return True

        return False




graph = Graph()

graph.add_edge(0,1)
graph.add_edge(1,2)
graph.add_edge(1,3)
graph.add_edge(1,4)
graph.add_edge(2,3)
graph.add_edge(2,4)
graph.add_edge(3,4)


g2 = Graph()

g2.add_edge(1,2)
g2.add_edge(2,3)
g2.add_edge(4,5)
g2.add_edge(5,6)
g2.add_edge(8,9)

# graph.bfs(3)
print(graph.has_cycle())
print(g2.has_cycle())
# print(graph.bfs(3))